### DISCUSSION TITLE
Character rotation mode issues with movement base rotation

### DISCUSSION THREAD

**Discussion by DarkiRuiz** [Sixze/ALS-Refactored#254] [General]
I think the ALS character behaves incorrectly when standing on rotating objects, like the  the spinning platform in the L_Playground level. By "movement base rotation" I'm using the engine's own definition, which is the rotation of whatever the character is standing on.

In case of VelocityDirection mode, I expect the character to "spin" along with the platform, keeping their rotation relative to the platform unchanged. The actual behavior is that they constantly rotate in place trying to maintain last input direction. I assume that the issue here is that the character isn't offsetting the calculated velocity direction by the base's delta rotation.

For LookingDirection mode, the behavior seems correct in this case (after all we want the character to maintain the camera look rotation), but the issue here (in my opinion) is that the turn in place action isn't triggering even when the difference between look direction and velocity direction is big enough. I assume this happens because the constant rotate in place action keeps the turn in place action from triggering.

Aiming mode seems to work correctly to me.

A related consideration would be first person view mode. I think this is a bit more subjective because I see valid reasons to support both options (to follow the base rotation or not). Maybe it should ignore the base rotation only when aiming? In any case, maybe having a direct method to toggle this would be a good idea.

**Comment by ameaninglessname** [comment]
https://github.com/Sixze/ALS-Refactored/blob/6039816921270fd4503e7f896a7796093a3d6bd8/Source/ALS/Private/AlsAnimationInstance.cpp#L901

Just looked around a bit, looks like these feet target locations need to be relative location to the "movement base".

**Comment by DarkiRuiz** [comment]
Hmm... I guess the solution will involve checking how that part goes, but I'd say that the feet are sorta working "properly", in the sense that since the character is trying to keep their rotation fixed while the floor moves, the feet should be doing that, otherwise they'd be "sliding".

What I mean to say is that it's the whole character what seems to be ignoring the base's rotation, not the feet.
Could it be related to the controller's rotation? ALS uses it to figure out the character's direction, but since the controller isn't really "standing" there, I would assume that stuff like spinning floors wouldn't affect it by default.

I remember that I had similar issues while using the marketplace version of ALS in a prior project some years ago.
Iirc what I did was to keep track of any changes in the "movement base" at the movement comp, and if the thing wasn't static, I would calculate a delta rotation so all I had to do if necessary was to add that to the characters rotation or anything related.

**Comment by ameaninglessname** [comment]
Yeah, make sense.

> In case of VelocityDirection mode, I expect the character to "spin" along with the platform, keeping their rotation relative to the platform unchanged.
Agreed with this. By looking at the code, when the difference of the foot location is large enough with target location, it would `PlayQueuedDynamicTransitionAnimation`.
https://github.com/Sixze/ALS-Refactored/blob/6039816921270fd4503e7f896a7796093a3d6bd8/Source/ALS/Private/AlsAnimationInstance.cpp#L1352-L1354

And because the rotating base is keeping moving, so the target location of the foot also changes,
then its playing the transitions forever.

Indeed, if we take consideration of the "movement base", looks like lots of location rotation stuff would need to be relative to it.

**Comment by Sixze** [comment]
I tried to implement something like this in the past, but I didn't succeed. I don't remember exactly what went wrong, but it was probably something network related, like the character started jittering or something like that.

I'll try to deal with this issue again in the future, but I can't promise much.

**Comment by Sixze** [comment]
I've added proper support for the `UCharacterMovementComponent::bIgnoreBaseRotation` setting, now it works as expected, when set to `false` the character and camera will rotate relative to the movement base.

I also added the `UAlsCharacterSettings::bInheritMovementBaseRotationInVelocityDirectionRotationMode` setting, which makes the character rotate with the movement base in velocity direction rotation mode, even when `bIgnoreBaseRotation` is set to `true`.

Personally, I recommend leaving `bIgnoreBaseRotation` set to `true` and only using `bInheritMovementBaseRotationInVelocityDirectionRotationMode` if necessary as `UCharacterMovementComponent` by itself does not support based movement very well, most calculations still happen in world coordinates, and in multiplayer things are even worse.

**Reply by epcau** [reply]
Out of curiosity, what kind of issues arise when bInheritMovementBaseRotationInVelocityDirectionRotationMode is left on?
I have tried enabling it and having a mesh rotating and moving around, I couldn't find any immediate issues using the character, neither on multiplayer (didn't try while simulating network latency though)
